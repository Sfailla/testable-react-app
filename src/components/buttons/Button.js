import React from 'react';
import { createUseStyles } from 'react-jss';
import PropTypes from 'prop-types';

const useStyles = createUseStyles(theme => ({
	buttonGroup: {
		width: '100%',
		display: 'flex',
		justifyContent: 'space-between',
		paddingTop: theme.spacing(5)
	},
	square: {
		border: 'none',
		outline: 'none',
		fontWeight: 'bold',
		letterSpacing: '3px',
		textTransform: 'uppercase'
	}
}));

function Button({ text, variant, className, ...otherProps }) {
	const classes = useStyles();
	const withClassName = `${classes[variant]} ${className || ''}`;

	return (
		<button className={withClassName} {...otherProps}>
			{text}
		</button>
	);
}

Button.propTypes = {
	text: PropTypes.string.isRequired,
	variant: PropTypes.string,
	onClick: PropTypes.func
};

function ButtonGroup({ children }) {
	const classes = useStyles();
	return <div className={classes.buttonGroup}>{children}</div>;
}

ButtonGroup.propTypes = {
	children: PropTypes.array
};

export { Button, ButtonGroup };
