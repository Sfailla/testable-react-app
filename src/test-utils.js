import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'react-jss';
import { theme } from './styles/theme';

const AllMyProviders = ({ children }) => {
	return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const MyCustomRender = (component, options) =>
	render(component, { wrapper: AllMyProviders, ...options });

// re-export everything from the `react-testing-library`
export * from '@testing-library/react';

// export our custom render method
export { MyCustomRender as render };
