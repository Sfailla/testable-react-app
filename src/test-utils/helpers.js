import React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'react-jss';
import { theme } from '../styles/theme';
import { StaticRouter } from 'react-router-dom';

function shallowWithTheme(child) {
	return shallow(child, {
		wrappingComponent: ({ children }) => (
			<ThemeProvider theme={theme}>{children}</ThemeProvider>
		)
	});
}

function mountWithTheme(child) {
	return mount(child, {
		wrappingComponent: ({ children }) => (
			<ThemeProvider theme={theme}>{children}</ThemeProvider>
		)
	});
}

function wrapWithRouter({ children }) {
	return <StaticRouter>{children}</StaticRouter>;
}

module.exports = {
	shallowWithTheme,
	mountWithTheme,
	wrapWithRouter
};
