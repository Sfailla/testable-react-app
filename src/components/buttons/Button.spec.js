import React from 'react';
import { mountWithTheme } from '../../test-utils/helpers';
import { render, fireEvent } from '../../test-utils';

import { Button, ButtonGroup } from './Button';

const mockFn = jest.fn();
const props = {
	text: 'click me',
	variant: 'round',
	onClick: mockFn
};

// testing with ENZYME
describe('<Button /> unit tests', () => {
	it('should render Button 1 time', () => {
		const wrapper = mountWithTheme(<Button {...props} />);
		expect(wrapper.find('Button')).toHaveLength(1);
	});

	it('<Button /> should contain button element', () => {
		const wrapper = mountWithTheme(<Button {...props} />);
		expect(wrapper.find('button')).toHaveLength(1);
	});

	it('should change text prop from original value to a new value', () => {
		const wrapper = mountWithTheme(<Button {...props} />);
		wrapper.setProps({ text: 'go back' });
		expect(wrapper.props().text).toEqual('go back');
	});

	it('should call mockFn when clicked', () => {
		const wrapper = mountWithTheme(<Button {...props} />);
		wrapper.find('Button').simulate('click');
		expect(mockFn).toHaveBeenCalled();
	});
});

describe('<ButtonGroup /> unit tests', () => {
	it('should render <ButtonGroup /> 1 time', () => {
		const wrapper = mountWithTheme(<ButtonGroup />);
		expect(wrapper.find('ButtonGroup')).toHaveLength(1);
	});

	it('<ButtonGroup /> should contain <div> tag', () => {
		const wrapper = mountWithTheme(<ButtonGroup />);
		expect(wrapper.find('div')).toHaveLength(1);
	});

	it('should match snapshot', () => {
		const wrapper = mountWithTheme(<ButtonGroup />);
		expect(wrapper).toMatchSnapshot();
	});
});

// Testing with TESTING LIBRARY
describe('<Button /> unit test', () => {
	// it is recommended to search for button using screen.getByRole query.  this is done by
	// screen.getByRole('button', {name: /submit/i})

	it('should render <Button /> component without errors', () => {
		render(<Button {...props} />);
	});

	it('Button text value should be correct', () => {
		const { queryByText } = render(<Button {...props} />);
		expect(queryByText('click me')).toBeTruthy();
	});

	it('text value of Button should change', () => {
		const { queryByText, rerender } = render(<Button {...props} />);
		expect(queryByText('click me')).toBeTruthy();

		const textValue = 'go back';
		rerender(<Button {...props} text={textValue} />);
		expect(queryByText(textValue)).toBeTruthy();
	});

	it('should call mockFn when clicked', () => {
		const { onClick } = props;
		const { getByText } = render(
			<Button {...props} onClick={onClick} />
		);
		fireEvent.click(getByText(props.text));
		expect(onClick).toHaveBeenCalled();
	});

	it('should match snapshot', () => {
		const { container } = render(<Button {...props} />);
		expect(container).toMatchSnapshot();
	});
});

describe('<ButtonGroup /> unit tests', () => {
	it('should render <ButtonGroup /> correctly', () => {
		render(<ButtonGroup />);
	});

	it('should match snapshot', () => {
		const { container } = render(<ButtonGroup />);
		expect(container).toMatchSnapshot();
	});
});
