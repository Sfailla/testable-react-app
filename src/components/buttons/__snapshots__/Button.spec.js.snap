// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> unit test should match snapshot 1`] = `
<div>
  <button>
    click me
  </button>
</div>
`;

exports[`<ButtonGroup /> unit tests should match snapshot 1`] = `
<ButtonGroup>
  <div
    className="buttonGroup-0-2-1"
  />
</ButtonGroup>
`;

exports[`<ButtonGroup /> unit tests should match snapshot 2`] = `
<div>
  <div
    class="buttonGroup-0-2-1"
  />
</div>
`;
